packages/async-chunks/src/lib/components/AsyncChunk.ts

```diff
@@ -0,0 +1,210 @@
+
+    constructor(props) {
+      super(props);
+      init();
+    }
+
+    componentWillMount() {
+      this.mounted = true;
+      this.loadModule();
+    }
+
+    componentWillUnmount() {
+      this.mounted = false;
+      this.clearTimeouts();
+    }
+
+    render() {
+      const {loading, error, pastDelay, timedOut, loadedChunk} = this.state;
+      if (loading || error) {
+        return React.createElement(opts.loading as any, {
+          isLoading: loading,
+          pastDelay,
+          timedOut,
+          error,
+          retry: this.retry,
+        });
+      } else if (loadedChunk) {
+        return opts.render && opts.render(loadedChunk, this.props);
+      } else {
+        return null;
+      }
+    }
+
+    loadModule() {
+      if (this.context.loadable && Array.isArray(opts.modules)) {
+        // report SSR modules with the module name (i.e the async chunks detected)
+        opts.modules.forEach(moduleName => {
+          this.context.loadable.report(moduleName);
+        });
+      }
+
+      if (chunk && !chunk.loading) {
+        return;
+      }
+
+      if (typeof opts.delay === 'number') {
+        if (opts.delay === 0) {
+          this.setState({pastDelay: true});
+        } else {
+          this.delay = setTimeout(() => {
+            this.setState({pastDelay: true});
+          }, opts.delay);
+        }
+      }
+
+      if (typeof opts.timeout === 'number') {
+        this.timeout = setTimeout(() => {
+          this.setState({timedOut: true});
+        }, opts.timeout);
+      }
+
+      const promise = chunk && chunk.promise;
+      promise &&
+        promise
+          // eslint-disable-next-line
+          .then(() => {
+            this.update();
+          })
+          .catch(err => {
+            this.update();
+            throw err;
+          });
+    }
+
+    update() {
+      if (!this.mounted) {
+        return;
+      }
+
+      this.setState({
+        error: chunk && chunk.error,
+        loadedChunk: chunk && chunk.loadedChunk,
+        loading: (chunk && chunk.loading) || false,
+      });
+
+      this.clearTimeouts();
+    }
+
+    retry() {
+      this.setState({error: null, loading: true, timedOut: false});
+      chunk = chunkLoader(opts.loader);
+      this.loadModule();
+    }
+
+    clearTimeouts() {
+      clearTimeout(this.delay);
+      clearTimeout(this.timeout);
+    }
+  };
+}
+
+export function preloadAll() {
+  return new Promise((resolve, reject) => {
+    // eslint-disable-next-line promise/catch-or-return
+    flushInitializers(ALL_INITIALIZERS).then(resolve, reject);
+  });
+}
+
+export function preloadReady() {
+  return new Promise(resolve => {
+    // Will always resolve, errors should be handled within loading UIs.
+    // eslint-disable-next-line promise/catch-or-return
+    flushInitializers(READY_INITIALIZERS).then(resolve, resolve);
+  });
+}
+
+export default function asyncChunk(opts) {
+  return asyncChunkGenerator(chunkLoader, opts);
+}
```

Comment:
OK, so I imagine this mostly just came from loadable, but IMO it does what it does in a bad way. Here is the API I think would be more appropriate:

```ts
// This will be used to collect the async chunks, and to extract them at the end for the
// HTML render.
class AsyncComponentManager {}

// This just provides the async manager to the rest of the tree
export class AsyncComponentProvider extends React.Component<{manager: AsyncComponentManager}> {}

// This component manages the actual async loading. It is roughly equivalent to the
// LoadableComponent here, but can operate independently from the factory function,
// so that someone can easily do <AsyncComponent loader={() => import('./Foo')} />
// anywhere in the app, without needing to construct a special component for it. It would use
// AsyncComponentManager for managing the chunks, and ideally would also use something
// like [this](https://github.com/lemonmade/perfect-app/blob/master/packages/%40shopify/react-extract/server.ts) to use a method for SSR extraction that we could
// reuse in various other SSR things
export class AsyncComponent<AllTheLoaderErrorRenderEtcProps> {}

// A trivial wrapper function that just creates a new component that async loads the
// options passed in to it
export function createAsyncComponent(options: AllTheLoaderErrorRenderEtcProps) {
  return () => <AsyncComponent {...options} />;
}
```